src/java/com/zimbra/oauth/handlers/impl/FacebookContactsImport.java

```diff
@@ -0,0 +1,569 @@
+                refreshToken, authInfo.getClientId(), authInfo.getClientSecret(), encodedUrl);
+        final GetMethod request = new GetMethod(queryString);
+
+        JsonNode json = null;
+        try {
+            json = FacebookOAuth2Handler.executeRequestForJson(request);
+        } catch (final IOException e) {
+            ZimbraLog.extensions
+                        .errorQuietly("There was an issue acquiring the authorization code.", e);
+            throw ServiceException
+                        .PERM_DENIED("There was an issue acquiring an authorization code for this user.");
+        }
+        String code = null;
+        if (!json.has("error") && json.has("code")) {
+            code = json.get("code").asText();
+        } else {
+            ZimbraLog.extensions
+                    .errorQuietly("There was an issue acquiring the authorization code. Response: "
+                    + json.toString(), null);
+            throw ServiceException.PERM_DENIED("Required code from Facebook was not found.");
+        }
+
+        return code;
+    }
+
+    /**
+     * Requests contacts for the given credentials.
+     *
+     * @param url The contacts url
+     * @return Json contacts response
+     * @throws ServiceException If there are issues retrieving the data
+     * @throws IOException If there are issues executing the request
+     */
+    protected JsonNode getContactsRequest(String url) throws ServiceException, IOException {
+
+        final GetMethod get = new GetMethod(url);
+        ZimbraLog.extensions.trace("Fetching contacts for import.");
+        return OAuth2Handler.executeRequestForJson(get);
+    }
+
+    @Override
+    public void importData(List<Integer> folderIds, boolean fullSync) throws ServiceException {
+        final String refreshToken = refresh();
+        final Mailbox mailbox = mDataSource.getMailbox();
+        final int folderId = mDataSource.getFolderId();
+        // existing contacts from the datasource folder
+        final Set<String> existingContacts = getExistingContacts(mailbox,
+                folderId, A_imAddress1);
+
+        String respContent = "";
+        String nextPageUrl = null;
+        try {
+            do {
+                // build contacts url, query params with access token or use Facebook nextPageUrl if defined
+                final String url = buildContactsUrl(nextPageUrl, refreshToken);
+                nextPageUrl = null;
+                // log this only at the most verbose level, because this contains
+                // privileged information
+                ZimbraLog.extensions.debug(
+                        "Attempting to sync Facebook contacts. URL: %s", url);
+                final JsonNode jsonResponse = getContactsRequest(url);
+                respContent = jsonResponse.toString();
+                if (jsonResponse.has("error") && jsonResponse.get("error").isContainerNode()) {
```

Comment:
You check for `jsonResponse != null` in a few places - wrap it all in one `if`? This one is missing a null check. Don't check if `get("error").isContainerNode`.